### ISSUE TITLE
What do we need to integrate with the clipboard api?

### ISSUE THREAD

**Issue by johanneswilm** [w3c/editing#64]
contentEditable=Events lacks execCommand, so what do we need in order to still allow clipboard access (using custom UI controls)?

**Comment by hallvors** [comment]
I think the answer is "nothing"?

Certainly there's still a method called execCommand on document if there is an element with cE=events in the DOM. Calling `execCommand('copy')` should work document-wide, including any selection that might be inside a cE=events element. The copy event will fire so a script that has registered a 'copy' listener (for example on document) will be able to control what is written to the clipboard.

At some point in the future we make a better clipboard API (say `window.clipboard.add(clipboardDataItem) .. window.clipboard.commit()`) and there's still nothing you need to do in your spec, because script authors will just use those commands directly, and it will just work.

Right?

**Comment by johanneswilm** [comment]
The spec is meant to not let CE Event Editing Hosts work as editing hosts for execCommand. So as far as commands that don't need an editing host goes, I guess they should still function. For those that do require an editing host, they will not work. If I understand you right, the copy command should not require an editing host, so in theory that should still work.

How about paste and cut?

**Comment by johanneswilm** [comment]
The spec is meant to not let CE Event Editing Hosts work as editing hosts for execCommand. So as far as commands that don't need an editing host goes, I guess they should still function. For those that do require an editing host, they will not work. If I understand you right, the copy command should not require an editing host, so in theory that should still work. Cut should be implementable through the copy logic.

Paste should not work at all through custom UI buttons, If I understand you right.

**Comment by kojiishi** [comment]
Maybe I'm worried a bit too much, but Blink (and probably WebKit too) canonicalize selections when set as of today, as you all know. CE=events is supposed to stop that, correct?

All existing command implementations in Blink assume that selections are canonicalized before executions. That said, there're quite a risk where those commands do not work against non-canonicalized selections.

An easy fix from top of my head is to canonicalize the current selection in the beginning of `execCommand()`. That could alter the selection even for `'copy'`. Would that be acceptable? If so, it might be nice to mention that in the spec.

If no, hm, this could be a hard problem to solve. Thoughts?

**Comment by johanneswilm** [comment]
> Maybe I'm worried a bit too much, but Blink (and probably WebKit too) canonicalize selections when set as of today, as you all know.

I was not aware of this, or I had forgotten about it. Do you mean it manipulates the DOM (merging sibling text nodes, removing empty text nodes with no siblings, etc.) just because something is selected or copied? That, to me, would not be good. Especially for collaborative editors this would be a big problem.

If, on the other hand, the contents of the clipboard do not entirely correspond to the content that was copied (because text nodes have been merged in the copied version, etc.) that would be much less of a problem.

Is there any reason why anything needs to be changed in the DOM just because it is being copied?

**Comment by johanneswilm** [comment]
Or do you mean that the selection is changed under such circumstances? And yes, that shouldn't happen with CE=events.

For the case of copying using execCommand -- instead of canonicalize the current selection before execCommand('copy'), could one copy the current selection and then do the canolicalization on the copied version without touching the current selection as it's available to the user?

**Comment by rniwa** [comment]
@kojiishi, canonicalization can be moved to the beginning of each editing command.  As far as copy is concerned, that command canonicalizes the selection within its implementation already so it shouldn't be an issue.  Even if it weren't the case, you could easily fix it.

**Comment by kojiishi** [comment]
@johanneswilm:

> Or do you mean that the selection is changed under such circumstances?

Yes.

> And yes, that shouldn't happen with CE=events.

If we allow `execCommand()` to work against CE=events, as @rniwa suggested, it's workable, but it's possible that Blink/WebKit move the canonicalizations to the top of `execCommand()` (or to the top of each command in `execCommand()`), so `execCommand()` is likely to change the selection. It'd be helpful if the spec mentions this; e.g. such as:

> DOM operations other than listed here may change the selections even if CE=events

If you have specific existing operations that you do not want to change the selections, it'd be helpful to list them in the spec.

**Comment by johanneswilm** [comment]
> If we allow execCommand() to work against CE=events

As I understand it, this will only be those operations that **do not** require an editing host (an editing host in the classic sense, we have now also changed the name of the CE=events container to include the term "editing host", but it shouldn't be an editing host for execCommand). And we only do it because otherwise we have no access at all to adding new items to the clipboard. So in those cases execCommand will treat CE=events as if it was any other, non-editable element.

> `execCommand()` is likely to change the selection

I have a feeling that changes to the selection for no good reason will be problematic for those who protested against these normalizations of the selection. But I cannot recall which of the editor people saw the normalization of selections as problematic. Was it one of you: @spocke,  @teleclimber or @fredck?

**Comment by Reinmar** [comment]
> I have a feeling that changes to the selection for no good reason will be problematic for those who protested against these normalizations of the selection. But I cannot recall which of the editor people saw the normalization of selections as problematic. Was it one of you: @spocke, @teleclimber or @fredck?

I'm pretty sure every single one of us saw it as problematic, but I recall that I was writing about this (even [yesterday](https://medium.com/content-uneditable/contenteditable-the-good-the-bad-and-the-ugly-261a38555e9c) actually). 

The most important problem with selection normalisation is that [it makes it impossible to place selection inside empty inline elements](https://bugs.webkit.org/show_bug.cgi?id=15256) which is crucial for implementing features like _bold_ and _italic_. User presses a _bold_ button, an editor creates `<b></b>` element, inserts it and tries to place the caret inside. 
- Gecko and IE will keep the selection inside that empty inline element. When user starts typing, the text will appear inside this element.
- Blink and WebKit moves the selection to the closest text node (so basically – somewhere outside). There's a small difference between Blink and WebKit currently, as Blink will claim that the selection is inside that element (until it's manually changed), while in WebKit `getRangeAt()` clearly shows that the selection was normalised. The [change in Blink](https://code.google.com/p/chromium/issues/detail?id=2874), as I recall, was made as a first step to fix this bug, but the second step, [fixing commands so they don't normalise selection](https://code.google.com/p/chromium/issues/detail?id=346613) wasn't made.

Note that setting selection inside an empty inline element isn't the only problematic case. There are more:
- `<b>foo</b>^<b>bar</b>` (can be achieved by removing bold from a collapsed selection)
- `<p>^<b>foo</b>` (can be achieved by removing bold from a collapsed selection)

Interesting fact: if you use `document.execCommand( 'bold' )`, then magically everything works fine, but not because the engine allows to set a selection in such a place, but because the inline element is created upon text insertion. While this can be regarded as a solution (that an editor records that the _bold_ command was executed and creates an element later on) I think it's not a good direction. For instance, the style of the caret (weight and colour of the line and cursive) could one day reflect the inline styling (it partially works in MSWord and very nicely in Google Docs). As a user I would then expect that if I press the _bold_ button the caret immediately changes. That could be triggered by setting selection inside that `<b>` or `<i>` element. Therefore, I prefer Gecko's and IE's behaviour.

Moreover, when using `contentEditable=true` working around this issue in Blink and WebKit requires terrible hacks (CKEditor uses zero-with-space which it later must remove without breaking a composition). So in my opinion this problem needs to be resolved anyway and aligning specification to that flawed Blink's and WebKit's behaviour seems strange.

**Comment by Reinmar** [comment]
PS. A thing I don't understand – why canonicalisation must touch selection. Cannot it be done on a copy of ranges? The commands algorithms could still expect to work on a normalised state, while executing a command wouldn't touch the real selection. Or are those algorithms to closely tied to a selection?

**Comment by Reinmar** [comment]
> The spec is meant to not let CE Event Editing Hosts work as editing hosts for execCommand. So as far as commands that don't need an editing host goes, I guess they should still function. For those that do require an editing host, they will not work. If I understand you right, the copy command should not require an editing host, so in theory that should still work. Cut should be implementable through the copy logic.

It seems so.

> Paste should not work at all through custom UI buttons, If I understand you right.

Why shouldn't paste work? It works right now in `contentEditable=true` if `execCommand()` was executed in a trusted event.

**Comment by johanneswilm** [comment]
@Reinmar Thanks for the reminder! Now the copy event only makes sense if the selection isn't collapsed, right? So in theory, if they would simply cancel the normalization/event of copy if the selection is collapsed, those points would be covered, right?

Can you recall whether normalization of non-collapsed selection also gives problems? In general I would agree that they they should make some kind of browser internal copy of the selection and then do everything on that, but I don't know the browser internals either.

> > Paste should not work at all through custom UI buttons, If I understand you right.
> 
> Why shouldn't paste work? It works right now in contentEditable=true if execCommand() was executed in a trusted event.

I am not sure what considerations went into that, but this is how I interpret this passage:

> Scripts can not trigger paste actions, not even from user-initiated threads (though the spec leaves the door open to implementing specific permission UIs that will allow this on a per-site basis).

http://www.whatcouldbewrong.com/articles/html5-clipboard-support/

**Comment by spocke** [comment]
I agree with @Reinmar having the selection canonicalized makes it hard to control selection.

If we want to keep execCommand("cut/copy") and use them for cE=events then they shouldn't mutate the selection otherwise one would have to do ugly workarounds to restore the selection after a cut/copy call if you want it to be in a specific place.

So I guess the spec needs to have some details on how cut/copy should work in a cE=events environment also I think it should be the only commands available until an replacement API to clipboard access is spec:ed.

**Comment by johanneswilm** [comment]
@spocke Do we need execCommand("cut")? Wouldn't it be enough to do execCommand("copy")?

**Comment by spocke** [comment]
Yeah, that true since if it doesn't do anything to the editing host by default.

**Comment by Reinmar** [comment]
> I am not sure what considerations went into that, but this is how I interpret this passage:
> 
> > Scripts can not trigger paste actions, not even from user-initiated threads (though the spec leaves the door open to implementing specific permission UIs that will allow this on a per-site basis).
> 
> http://www.whatcouldbewrong.com/articles/html5-clipboard-support/

Thanks for the link. You can ignore my previous comment because I incorrectly remembered that Chrome and FF will/do support user-initiated paste, while they only support paste and cut. Hallvord's post makes it super clear.

I was wondering though, if the spec leaves the door open for user-initiated paste, what if I change browser's settings to allow clipboard access (if any implements this). Hypothetically, the `paste` command would work then in `contentEditable=true` but what would happen in `contentEditable=events`? AFAICS we don't need to spec it at all, because the `paste` event will be fired on the document, so it's the same case as with `copy`. It means that whatever happens `cE=events` should be safe (as @hallvors wrote).

> Yeah, that true since if it doesn't do anything to the editing host by default.

Agree.

**Comment by hallvors** [comment]
> AFAICS we don't need to spec it at all, because the paste event will be fired on the document

Indeed. The paste event will fire on the document, script can handle data from the clipboard and insert as desired. The clipboard spec says about the paste action:

> If the event was not canceled
>    If there is a selection or cursor in an editable context where pasting is enabled

Note _where pasting is enabled_ - this was added to let you say pasting is not enabled for a cE=events element, and thus the paste action will not insert anything by itself.

**Comment by johanneswilm** [comment]
> Note where pasting is enabled - this was added to let you say pasting is not enabled for a cE=events element, and thus the paste action will not insert anything by itself.

Right, we coordinated this to make sure this wouldn't apply in the events-spec some months ago.

The only thing left, as I can see it then is to:

-- Possibly make it even clearer that the CE=events container does not work for execCommand commands that require an editing host
-- Make it clear that copying content does not lead to the selection being affected in any way.

Are there any other execCommand commands that do not require an editing host (other than copy)? Do we need to specify that these should be disabled as well?

**Comment by hallvors** [comment]
There are traces of documentation of commands like "SaveAs" and "Bookmark" on the web, which may not have required an editing host - but I don't think you need to worry about those obscurities ;)

Note that I also intend to "disconnect" cut and paste from editing hosts - if the script can handle a paste it should not have to resort to hacks like hidden TEXTAREAs to enable pasting. But as far as I can tell, cE=events doesn't need to worry about these details either. A script developer adding 

```
document.addEventListener('beforepaste', function(e){e.preventDefault()})
```

has enabled the 'paste' menu entry and ctrl+V and will be able to get data from the paste event even if there is no editing host or the focus is in a cE=events element where the UA won't insert paste data.
